common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -452,22 +489,69 @@ void removeAppShuffleInfoFromDB(AppShuffleInfo appShuffleInfo) {
+      AppShuffleInfo shuffleInfo, Integer shuffleId,
+      Optional<AppAttemptShuffleMergeId> higherShuffleMergeId) {
+    AppShuffleMergePartitionsInfo mergePartitionsInfo = shuffleInfo.shuffles.get(shuffleId);
+    AppAttemptShuffleMergeId shuffleMergeId = new AppAttemptShuffleMergeId(
+        shuffleInfo.appId, shuffleInfo.attemptId, shuffleId, mergePartitionsInfo.shuffleMergeId);
+    if (!mergePartitionsInfo.isFinalized()) {
+      Map<Integer, AppShufflePartitionInfo> partitionsToClean = mergePartitionsInfo.shuffleMergePartitions;
+      submitCleanupTask(() ->
+          closeAndDeleteOutdatedPartitions(shuffleMergeId, partitionsToClean, higherShuffleMergeId));
+    } else {
+      int[] partitionsToClean = mergePartitionsInfo.finalizedPartitions;
+      submitCleanupTask(() ->
+          deleteOutdatedFinalizedPartitions(shuffleInfo, shuffleMergeId, partitionsToClean, higherShuffleMergeId));
+    }
+  }
+
   @VisibleForTesting
   void closeAndDeleteOutdatedPartitions(
-      AppAttemptShuffleMergeId appAttemptShuffleMergeId,
-      Map<Integer, AppShufflePartitionInfo> partitions) {
-    removeAppShufflePartitionInfoFromDB(appAttemptShuffleMergeId);
+      AppAttemptShuffleMergeId shuffleMergeId,
+      Map<Integer, AppShufflePartitionInfo> partitions,
+      Optional<AppAttemptShuffleMergeId> higherShuffleMergeId) {
+    removeAppShufflePartitionInfoFromDB(shuffleMergeId);
+    higherShuffleMergeId.ifPresent(this::removeAppShufflePartitionInfoFromDB);
     partitions
-      .forEach((partitionId, partitionInfo) -> {
-        synchronized (partitionInfo) {
-          partitionInfo.closeAllFilesAndDeleteIfNeeded(true);
-        }
-      });
+        .forEach((partitionId, partitionInfo) -> {
+          synchronized (partitionInfo) {
+            partitionInfo.closeAllFilesAndDeleteIfNeeded(true);
+          }
+        });
+  }
+
+  @VisibleForTesting
+  void deleteOutdatedFinalizedPartitions(
+      AppShuffleInfo shuffleInfo,
+      AppAttemptShuffleMergeId shuffleMergeId,
+      int[] outdatedFinalizedPartitions,
+      Optional<AppAttemptShuffleMergeId> higherShuffleMergeId) {
+    int shuffleId = shuffleMergeId.shuffleId;
+    int mergeId = shuffleMergeId.shuffleMergeId;
+    removeAppShufflePartitionInfoFromDB(shuffleMergeId);
+    higherShuffleMergeId.ifPresent(this::removeAppShufflePartitionInfoFromDB);
+    Arrays.stream(outdatedFinalizedPartitions).forEach(partition -> {
+      try {
+        File metaFile =
+            shuffleInfo.getMergedShuffleMetaFile(shuffleId, mergeId, partition);
+        File indexFile = new File(
+            shuffleInfo.getMergedShuffleIndexFilePath(shuffleId, mergeId, partition));
+        File dataFile =
+            shuffleInfo.getMergedShuffleDataFile(shuffleId, mergeId, partition);
+        metaFile.delete();
+        indexFile.delete();
+        dataFile.delete();
+      } catch (Exception e) {
+        logger.error("Error delete shuffle files for {}", shuffleMergeId, e);
+      }
```

Comment:
Just like `closeAllFilesAndDeleteIfNeeded` method, can we continue delete the other files if one `delete()` failed?

Reply:
Thanks for your review~ done